### PR TITLE
Remove many build options redundantly specified in Python build scripts.

### PR DESCRIPTION
**Pull request by xuhdev**
Currently many build options are explicitly passed from Python build scripts to CMake. But this is unecessary, at least for many of them. This commit removes the build options that have the same name in CMakeLists.txt and environment variables (e.g., `USE_REDIS`). Additionally, many build options that are not explicitly passed to CMake are lost.

For `ONNX_ML`, `ONNX_NAMESPACE`, and `BUILDING_WITH_TORCH_LIBS`, I changed their default values in CMake scripts (as consistent with what the `CMake.defines` call meant), to avoid their default values being redundantly set in the Python build scripts.

### CODE REVIEW

**Comment by xuhdev** [comment]
Note that the failed lint checking is unrelated to this PR

File: tools/setup_helpers/cmake.py [hunk 1]
```diff
@@ -232,61 +232,73 @@ def generate(self, version, cmake_python_library, build_python, build_test, my_e
...
+            'BUILDING_WITH_TORCH_LIBS',
+            'CMAKE_PREFIX_PATH',
+            'ONNX_ML',
+            'ONNX_NAMESPACE',
+            'WERROR'
+        }
+        for var, val in my_env.items():
+            # We currently pass over all environment variables that start with "BUILD_" or "USE_". This is because we
+            # currently have no reliable way to get the list of all build options we have specified in CMakeLists.txt.
+            # (`cmake -L` won't print dependent options when the dependency condition is not met.) We will possibly
```
Comment (ezyang) [inline tools/setup_helpers/cmake.py]: Alternately, does that mean we should just get rid of dependent options?

Comment (ezyang) [reply tools/setup_helpers/cmake.py]: Some more thinking: it's very useful to have some way of telling the user what all available options / environment variables are. Right now we just hand document a few notable ones in `setup.py` but if there were the moral equivalent of `--help` that seems useful for users. (Nothing to do for this PR)

**Review by ezyang — APPROVED** [review]

File: tools/setup_helpers/cmake.py (hunk 1 above)
Comment (xuhdev) [reply tools/setup_helpers/cmake.py]: I am thinking about having all option descriptions in `CMakeLists.txt` in standard CMake way, and creating a method in a Python script extracting these descriptions. So long as we restrict ourselves to write these options (and their descriptions and dependency relationship) in a way that the to-be-created Python method can recognize (e.g., no super-convoluted CMake syntax playing), and I do not see a problem.

**Comment by ezyang** [comment]
OK sgtm

Excerpts from Hong Xu's message of 2019-06-24 10:52:22 -0700:
> xuhdev commented on this pull request.
> 
> > +            # The default value cannot be easily obtained in CMakeLists.txt. We set it here.
> +            'CMAKE_PREFIX_PATH': distutils.sysconfig.get_python_lib()
> +        }
> +        # Options that do not start with 'USE_' or 'BUILD_' and are directly controlled by env vars
> +        additional_options = {
> +            'BLAS',
> +            'BUILDING_WITH_TORCH_LIBS',
> +            'CMAKE_PREFIX_PATH',
> +            'ONNX_ML',
> +            'ONNX_NAMESPACE',
> +            'WERROR'
> +        }
> +        for var, val in my_env.items():
> +            # We currently pass over all environment variables that start with "BUILD_" or "USE_". This is because we
> +            # currently have no reliable way to get the list of all build options we have specified in CMakeLists.txt.
> +            # (`cmake -L` won't print dependent options when the dependency condition is not met.) We will possibly
> 
> I am thinking about having all option descriptions in `CMakeLists.txt` in standard CMake way, and creating a method in a Python script extracting these descriptions. So long as we restrict ourselves to write these options (and their descriptions and dependency relationship) in a way that the to-be-created Python method can recognize (e.g., no super-convoluted CMake syntax playing), and I do not see a problem.
>

**Pull request closed without merging**